result.go

```diff
@@ -0,0 +1,220 @@
+}
+
+// Data decodes result data into out.
+func (e *Result) Data(out interface{}) error {
+	return json.Unmarshal([]byte(e.data), out)
+}
+
+// ResultEmitter is a MESG result event listener.
+type ResultEmitter struct {
+	app *Application
+
+	// resultTask is the actual event to listen for.
+	resultTask string
+
+	//resultServiceID is the service id of where result is emitted.
+	resultServiceID string
+
+	// outputKey is the output key to listen for.
+	outputKey string
+
+	// taskKey is the actual task that will be executed.
+	taskKey string
+
+	// executionTags keeps execution tags for filtering.
+	executionTags []string
+
+	// taskServiceID is the service id of target task.
+	taskServiceID string
+
+	// filterFuncs holds funcs that returns boolean values to decide
+	// if the task should be executed or not.
+	filterFuncs []func(*Result) bool
+
+	// mapFunc is a func that returns input data of task.
+	mapFunc func(*Result) Data
+
+	// executionTagsFunc is a func that returns execution tags filter.
+	executionTagsFunc func(*Result) []string
+
+	// gracefulWait will be in the done state when all processing
+	// results are done.
+	gracefulWait *sync.WaitGroup
+}
+
+// ResultCondition is the condition configurator for filtering results.
+type ResultCondition func(*ResultEmitter)
+
+// TaskKeyCondition returns a new option to filter results by task name.
+// Default is all(*).
+func TaskKeyCondition(task string) ResultCondition {
+	return func(e *ResultEmitter) {
+		e.resultTask = task
+	}
+}
+
+// OutputKeyCondition returns a new option to filter results by output key name.
+// Default is all(*).
+func OutputKeyCondition(key string) ResultCondition {
+	return func(e *ResultEmitter) {
+		e.outputKey = key
+	}
+}
+
+// TagsCondition returns a new option to filter results by execution tags.
+// This is a "match all" algorithm. All tags should be included in the execution to have a match.
+func TagsCondition(tags ...string) ResultCondition {
+	return func(e *ResultEmitter) {
+		e.executionTags = tags
+	}
+}
+
+// WhenResult creates a ResultEmitter for serviceID.
+func (a *Application) WhenResult(serviceID string, conditions ...ResultCondition) *ResultEmitter {
+	e := &ResultEmitter{
+		app:             a,
+		resultServiceID: serviceID,
+		resultTask:      "*",
+		outputKey:       "*",
+		gracefulWait:    &sync.WaitGroup{},
+	}
+	for _, condition := range conditions {
+		condition(e)
+	}
+	return e
+}
+
+// Filter sets filter funcs that will be executed to decide to execute the
+// task or not.
+// It's possible to add multiple filters by calling Filter multiple times.
+// Other filter funcs and the task execution will no proceed if a filter
+// func returns false.
+func (e *ResultEmitter) Filter(fn func(*Result) (execute bool)) *ResultEmitter {
+	e.filterFuncs = append(e.filterFuncs, fn)
+	return e
+}
+
+// SetTags sets execution tags for task executions.
+func (e *ResultEmitter) SetTags(fn func(*Result) (tags []string)) *ResultEmitter {
```

Comment:
@ilgooz I'm not ok with this function here.
It will be weird to use it. Eg:
```
WhenXXX()
.Filter()
.SetTags()
.Map()
.Execute()
```

What about adding option to the execution functions?

Anyway, can you remove it from this PR so we can merge and we discuss and implement it in a new issue/PR

Reply:
done